Introduce `lib.updateManyAttrsByPath`

infinisil opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Generalized https://github.com/NixOS/nixpkgs/pull/145952 to multiple updates and creation. I've also had the need for this before, leading me to implementing [modifyPaths](https://github.com/flox/flox-lib/blob/master/channel/utils/modifyPaths.nix) and [nestedListToAttrs](https://github.com/flox/flox-lib/blob/master/channel/utils/nestedListToAttrs.nix). The function in this PR is a generalization of both the PR and the two linked functions.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

infinisil committed:
lib.attrsets: Introduce updateManyAttrsByPath

Infinisil commented on lib/lists.nix:
```diff
@@ -226,6 +226,31 @@ rec {
+  #  if alen == lib.length b then compareElements 0
+  #  else alen < lib.length b;
+
+  lexicalLessThan =
```
See also https://github.com/NixOS/nix/pull/5631, which makes lists comparable in Nix directly

pennae commented:
this thing is really complex. could it be simplified with a modified comparison? without having checked it too far, this seems like it'd give the exact sort order you want to apply:
```
if isPrefix a b then false
else if isPrefix b a then true
else lexicalLessThan a b
```
assuming that the paths are short the first two conditions could be folded into a (then local and renamed) lexicalLessThan variant and probably not lose much performance

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-22/16251/1

Infinisil commented:
@pennae I believe that's pretty much what the `lexicalLessThan` does here, except that `isPrefix` is inlined (notably `isPrefix` doesn't currently exist for lists).

However this will become irrelevant soon, as https://github.com/NixOS/nix/issues/5627 has been merged, which makes this much faster anyways :)

pennae commented:
it shouldn't be the same, we'd want to sort `[ "a" "b" ]` *after* `[ "a" "b" "c" ]` (it'd be an easy change to to the nix version though by just flipping the booleans in the first two checks). that way you could run through the sorted list start to finish without further processing and still get the same order of applications you've specified for updateManyAttrPaths, but still have an easy optimization route of finding the end of operations to the current subtree with binary search (while nestedList generation is a linear time operation).

not sure how much difference it'd make in practice, we haven't benchmarked anything. making it more easily understandable does seem like it'd be very helpful though, your documentation makes it look like quite a monster right now :sweat_smile:

Infinisil commented:
@pennae Still not exactly sure what you mean. Here is a sorted list using the current ordering:
```
[
 [ ]
 [ "a" ]
 [ "a" "x" ]
 [ "a" "y" ]
 [ "b" ]
 [ "b" "x" ]
 [ "b" "y" ]
]
```

Do you mean to have an ordering like
```
[
 [ "a" "x" ]
 [ "a" "y" ]
 [ "a" ]
 [ "b" "x" ]
 [ "b" "y" ]
 [ "b" ]
 [ ]
]
```

instead? So like, sorting as if non-existing elements were larger than any existing ones (in comparison to lexicographic ordering, which treats non-existing elements as if they were smaller than any existing ones).

If so, I don't know how that would simplify anything, it would just inverse some operations

pennae commented:
that'd the be the order, yes. thought that might save you a search, but on closer inspection that doesn't seem to actually be the case. ignore everything i said :sweat_smile:

infinisil committed:
lib.attrsets: Introduce showAttrPath

infinisil committed:
lib.lists: Use builtins.groupBy for lib.groupBy

builtins.groupBy is much more performant. It was introduced in
https://github.com/NixOS/nix/pull/5715

Infinisil commented:
Update: Version 2 of this implementation (now in https://github.com/Infinisil/nixpkgs/tree/updateAttrPaths-v2) was very complex, but used way less memory than version 1, which was based on `lib.groupBy`. Unfortunately, despite implementing builtin lexical list comparison in Nix with https://github.com/NixOS/nix/issues/5627, version 2 was still a bit slower time-wise.

Considering that we need to change Nix to have better performance anyways, I revisited version 1 and noticed that if we had a `builtins.groupBy` it might turn out to work really well, so I made https://github.com/NixOS/nix/pull/5715 which does that. Turns out that indeed makes version 1 perform much better, both time-wise and memory-wise!

Here is a plot that shows the CPU time for each of these versions (sample size 100):
![plot](https://user-images.githubusercontent.com/20525370/144485000-fd55822f-6dfc-4110-b3f6-e7feb6b12ae9.png)

And here's a full comparison:
| | v1-lib | v2 | v1-lib |
| -- | -- | -- | -- |
| CPU time mean | 0.163s | 0.204s | 0.092s |
| Memory | 343.5MB | 97.3MB | 59.4MB |
| Nixpkgs commit | https://github.com/NixOS/nixpkgs/pull/147077/commits/ae4996c0d8b3a81dd8743efce79586f101fdd7cb | https://github.com/Infinisil/nixpkgs/commit/462a63a561296748cbda2ac6b202519d22c3a856 | https://github.com/NixOS/nixpkgs/pull/147077/commits/ae4996c0d8b3a81dd8743efce79586f101fdd7cb |
| Nix commit | https://github.com/nixos/nix/commit/2ff71b021379a2c9bbdcb789a93cdc585b3520ca | https://github.com/nixos/nix/commit/2ff71b021379a2c9bbdcb789a93cdc585b3520ca | https://github.com/NixOS/nix/pull/5715/commits/69d1518512f0a721c6a807613d9306bf35118e13 |

This was tested by evaluating the following file with `NIX_SHOW_STATS=1 nix-instantiate --eval file.nix`:
<details>

```nix
let
  lib = import ./lib;

  basicSet = n: builtins.listToAttrs (map (i: lib.nameValuePair (toString i) i) (lib.range 0 n));

  gen = n: if n == 0 then 0 else {
    l = gen (n - 1);
    r = gen (n - 1);
  };

  m = 12;

  input = gen m;

  genUpdates = n: if n == 0 then []
  else
    [{
      path = [];
      update = old: if builtins.isAttrs old then old // { changed = true; } else old + 1;
    }] ++
    map (u: u // { path = ["l"] ++ u.path; }) (genUpdates (n - 1)) ++
    map (u: u // { path = ["r"] ++ u.path; }) (genUpdates (n - 1));

  updates = genUpdates (m + 1);
  result = lib.attrsets.updateManyAttrPaths updates input;
in builtins.deepSeq result null
```
</details>

Infinisil commented:
I think this is ready now. I added a bunch of tests for this new function, but also for `groupBy` which was changed to use the builtin version in the future.

Infinisil commented:
I'm using ZFS as my filesystem, and I suspect @GrahamcOfBorg to also use ZFS. The build notably works if I use a tmpfs for the build directory!
```
$ sudo mkdir /tmpfs
$ sudo mount -t tmpfs -o size=4G,mode=1777 tmpfs /tmpfs
$ sudo TMPDIR=/tmpfs nix-build -A public-inbox
[...]
/nix/store/amxqkdv3ysg87j06anx3sly7ipyfsz6y-perl5.34.0-public-inbox-1.7.0
```

roberth commented on lib/attrsets.nix:
```diff
@@ -78,6 +78,87 @@ rec {
+            # each of them
+            if isAttrs input then
+              input //
+              mapAttrs (name: value: go (prefixLength + 1) value (input.${name} or {})) nested
```
Won't manage a full review today, but I did notice that you could rearrange the `go` params to support currying here, which the evaluator will like (even if it's minor).

Infinisil replied:
Applied this

roberth commented on lib/attrsets.nix:
```diff
@@ -78,6 +78,87 @@ rec {
+        # Recurses into the attribute set, passing the updates for each
+        # attribute respectively
+        withNestedMods =
+          if length split.wrong == 0 then
```
`split.wrong == []`

Save a call.

roberth commented on lib/attrsets.nix:
```diff
@@ -78,6 +78,87 @@ rec {
+      let
+        # Splits updates into ones on this level (split.right)
+        # and ones on levels further down (split.wrong)
+        split = partition (el: length el.path == prefixLength) updates;
```
This is not great if many updates have the same prefix. A `groupBy` based approach may be more efficient in such a case.

Infinisil replied:
I'm not sure what you mean by this.

Infinisil replied:
I don' think this is a problem, the predicate here is constant time.

roberth replied:
It's good enough. Could be optimized later if necessary.

roberth commented on lib/attrsets.nix:
```diff
@@ -78,6 +78,87 @@ rec {
+              throw
+              ( "updateManyAttrPaths: Path ${showAttrPath updatePath} needs to "
+              + "be updated, but path ${showAttrPath (take prefixLength updatePath)} "
+              + "of the given input is not an attribute set and therefore "
+              + "can't be recursed into");
```
```suggestion
              + "of the given input is not an attribute set, so we can't "
              + "update an attribute inside of it.");
```

roberth replied:
Though it's weird that missing ones are ignored.

Infinisil replied:
Updates to attributes that are missing are not ignored, or do you mean something else?

roberth commented on lib/tests/misc.nix:
```diff
@@ -761,4 +761,143 @@ runTests {
+      "a2-b"
+      "_bc'de"
+    ];
+    expected = "\".\".foo.\"2\".a2-b._bc'de";
```
`''` quotes are more readable.

roberth commented on lib/tests/misc.nix:
```diff
@@ -761,4 +761,143 @@ runTests {
+    };
+  };
+
+  # If an attribute doesn't exist, it is treated as if it did with a value of {}
```
This is too arbitrary for my taste at least.
How about adding `setWhenMissing` next to `update` to handle this case?

Infinisil replied:
The reason I chose `{}` is because this allows easily setting nested attributes:

```
nix-repl> :p lib.updateManyAttrPaths [ { path = [ "a" "b" "c" ]; update = _: 10; } ] {}   
{ a = { b = { c = 10; }; }; }
```

This works because when it tries to recurse into `a` and finds that there's no such attribute, it creates it with a value of `{}`. The same for the next path element.

This wouldn't work if e.g. `null` were used instead in the current implementation, it would result in an error:
```
nix-repl> lib.updateManyAttrPaths [ { path = [ "a" "b" "c" ]; update = old: 10; } ] {} 
{ a = «error: updateManyAttrPaths: Path a.b.c needs to be updated, but path a of the given input is not an attribute set, so we can't update an attribute inside of it.»; }
```

But after thinking about this again, this default of `{}` really only makes sense for initializing these intermediate path elements. For the last path element of an update, I very much like your suggestion of making it configurable. I'll try to implement this.

Infinisil replied:
I revisited this and determined that no such functionality is needed, because you can implement a fallback by just inserting another update. E.g. this example with your proposed `setWhenMissing`:

```nix
updateManyAttrPaths [
  {
    path = [ "a" ];
    update = old: old + "-updated";
    setWhenMissing = "missing";
  }
] {}
```

is equivalent to

```nix
updateManyAttrPaths [
  {
    path = [ "a" ];
    update = old: "missing";
  }
  {
    path = [ "a" ];
    update = old: old + "-updated";
  }
] {}
```

Infinisil replied:
However, I did remove the arbitrary default of `{}`, as it was only needed internally. Instead you will now get an error if you try to access an `old` argument when it doesn't exist.

roberth replied:
> ```nix
>     update = old: "missing";
> ```

Wouldn't that be destructive?

> ```nix
>     update = old: old + "-updated";
> ```

Or _if_ it wasn't destructive, wouldn't `old` still reference the exception?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-23/16631/1

Infinisil commented:
I think it's looking really good now. One thing I'm not yet sure about though is the name of this function. Which of these (or others) is best?
- `updateManyAttrPaths` (current name)
- `updateAttrPaths` (many is implicit from the `Paths`)
- `updateByAttrPaths` (but actually it's not updating the attribute paths, but rather the values at these paths, so add `By`)
- `updateAttrsByPaths` (but then we should probably mirror `lib.setAttrByPath` like this)

I feel like the last one is maybe the best

Profpatsch commented:
Hm, is this doing something that mergePatch can’t? I’ve got an implementation of that here: https://code.tvl.fyi/tree/nix/mergePatch/default.nix?id=4b5271e69851dcbf42794c3a1056a57b80a89041#n1

The advantages about mergePatch are that there is an RFC for it, so I’d expect it to be proven useful in practice.

Profpatsch commented:
From what I see updateManyAttrPaths can apply arbitrary update functions, while mergePatch only overrides/removes attrs, so it’s slightly more powerful.

In less serious terms, I had started to implement lenses in nix, which are a full superset of this :)
https://code.tvl.fyi/tree/users/Profpatsch/lens.nix?id=4b5271e69851dcbf42794c3a1056a57b80a89041#n106

But I’m not opposed to merging.

roberth commented:
> RFC

_IETF_ [RFC 7396](https://datatracker.ietf.org/doc/html/rfc7396). Interesting.

> so I’d expect it to be proven useful in practice.

I'm a teapot ;)
But yeah, there's a non-zero chance a Nix user may encounter it somewhere.

Both algorithms are useful.

In my experience working with functions like these, you want to use the least powerful function that satisfies your requirements, as it is generally the simplest and fastest. The latter is relevant for `lib` code like the module system.

So I'm also in favor of having both.

> https://code.tvl.fyi/tree/users/Profpatsch/lens.nix?id=4b5271e69851dcbf42794c3a1056a57b80a89041#n106

Profpatsch patching profunctors? :D quite a mouthful

Infinisil commented:
@Profpatsch Your `mergePatch` looks nice. There are some small differences I want to point out:
- `mergePatch` can't set fields to `null`, since it uses `null` to mean removal of an attribute
- Since `mergePatch` isn't based on modification functions, it's easy for information to be accidentally overridden
- `mergePatch` can only do static changes. It can't e.g. rename an attribute unless you access the previous value directly (outside of `mergePatch`)
- Accessing nested old attribute values outside of `mergePatch` is a bit inefficient, since you need to go through the attribute path again, e.g. the following needs to access `old.a.b.c` many times:
  ```nix
  let old = {
    a.b.c.x = 0;
    a.b.c.y = 1;
    a.b.c.z = 2;
  }; in mergePatch old {
    a.b.c.x = old.a.b.c.x + 1;
    a.b.c.y = old.a.b.c.y + 1;
    a.b.c.z = old.a.b.c.z + 1;
  }
  ```
- `mergePatch` is hard to use for dynamic updates, where the attribute paths to update are only determined through some processing. That is because the patch input to `mergePatch` is a nested attribute set, which is hard to process, since there's no way to indicate when the attribute path nesting ends and the data to process starts. This relates to `mapAttrsRecursiveCond`, which I think is pretty hard to understand and use properly

Major advantage to `mergePatch` is that it is much simpler to use for static updates (where it should also be faster), and that it has a proper specification backing it. So I think it would probably also be worthwhile to have it in nixpkgs, but I also think that it can't replace the function from this PR fully.

Profpatsch commented:
Yeah, especially the arbitrary functions for updates part is very nice to have. I’d say let’s merge this!

Infinisil commented:
@roberth @Profpatsch Nice! The only thing to discuss is the naming, see https://github.com/NixOS/nixpkgs/pull/147077#issuecomment-1057356279 I'm yet undecided

roberth commented on lib/attrsets.nix:
```diff
@@ -78,6 +78,103 @@ rec {
+       shallow attribute paths
+     - Multiple updates to the same attribute path are applied in the order
+       they appear in the update list
+     - If there is an update for an attribute path that would recurse into an
+       existing value that's not an attribute set, an error is thrown
```
"recurse" is part of the implementation, not the domain.

```suggestion
     - If any but the last `path` element leads into a value that is not an
       attribute set, an error is thrown.
```

roberth commented on lib/attrsets.nix:
```diff
@@ -477,6 +574,20 @@ rec {
+  */
+  showAttrPath = path:
+    if path == [] then "<root attribute path>"
+    else concatMapStringsSep "." escapeNixIdentifier path;
```
`showOption` looks hackier, but shouldn't these functions behave the same?

Infinisil replied:
See https://github.com/NixOS/nixpkgs/commit/124cccbe3b63122733e02e41e45a383ec48752fd

roberth commented on lib/tests/misc.nix:
```diff
@@ -761,4 +761,131 @@ runTests {
       { a = 3; b = 30; c = 300; }
     ];
   };
+
```
You could add the example from the docs here.

Infinisil replied:
Done

Infinisil commented:
In addition to addressing your review, I also made my own decision on the naming: `updateManyAttrsByPath`. This closely resembles the last option `updateAttrsByPaths` from my suggestions in https://github.com/NixOS/nixpkgs/pull/147077#issuecomment-1057356279, but I chose to keep the `Many` in there in order to distinguish it from a potential `updateAttrByPath` (which would be too close to `updateAttrsByPath`.
